### PR TITLE
[JENKINS-50030] Lazy-load-friendly `Job.getEstimatedDurationCandidates`

### PR DESCRIPTION
**Pull request by jglick**
Investigating how effective lazy-loading is, I noticed that running new builds often triggered loading of several old builds when this was not really necessary. With the API introduced here (need a follow-on PR in `workflow-job` to call it for Pipeline), `Job.getEstimatedDurationCandidates` will use a simpler algorithm which skips builds that were not already loaded from disk for some other reason.

### Testing done

Without the new test, old builds get loaded this way:

```
	at org.jvnet.hudson.test.RunLoadCounter$Marker.onLoad(RunLoadCounter.java:118)
	at hudson.model.Run.onLoad(Run.java:389)
	at hudson.model.RunMap.retrieve(RunMap.java:233)
	at hudson.model.RunMap.retrieve(RunMap.java:61)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:650)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:632)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:530)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.search(AbstractLazyLoadRunMap.java:495)
	at jenkins.model.lazy.LazyBuildMixIn$RunMixIn.getPreviousBuild(LazyBuildMixIn.java:400)
	at hudson.model.AbstractBuild.getPreviousBuild(AbstractBuild.java:198)
	at hudson.model.AbstractBuild.getPreviousBuild(AbstractBuild.java:106)
	at hudson.model.Job.getEstimatedDurationCandidates(Job.java:1040)
	at hudson.model.AbstractProject.getEstimatedDurationCandidates(AbstractProject.java:958)
	at hudson.model.Job.getEstimatedDuration(Job.java:1054)
	at hudson.model.queue.MappingWorksheet.<init>(MappingWorksheet.java:328)
	at hudson.model.queue.MappingWorksheet.<init>(MappingWorksheet.java:313)
	at hudson.model.Queue.maintain(Queue.java:1650)
```

Since this duration is indeed an “estimate”, it is not worth forcing builds to be loaded just for that purpose.

Preventing the `lastBuild` from being loaded is much harder:

```
	at org.jvnet.hudson.test.RunLoadCounter$Marker.onLoad(RunLoadCounter.java:120)
	at hudson.model.Run.onLoad(Run.java:389)
	at hudson.model.RunMap.retrieve(RunMap.java:233)
	at hudson.model.RunMap.retrieve(RunMap.java:61)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:650)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:632)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:530)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.search(AbstractLazyLoadRunMap.java:495)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.newestBuild(AbstractLazyLoadRunMap.java:428)
	at jenkins.model.lazy.LazyBuildMixIn.getLastBuild(LazyBuildMixIn.java:251)
	at hudson.model.AbstractProject.getLastBuild(AbstractProject.java:943)
	at hudson.model.AbstractProject.getLastBuild(AbstractProject.java:137)
	at hudson.model.Job.isLogUpdated(Job.java:314)
	at hudson.model.AbstractProject.getCauseOfBlockage(AbstractProject.java:1076)
	at hudson.model.Queue.getCauseOfBlockageForTask(Queue.java:1239)
	at hudson.model.Queue.getCauseOfBlockageForItem(Queue.java:1196)
	at hudson.model.Queue.maintain(Queue.java:1570)
```

when `concurrentBuild` is off (the default for non-Pipeline); when on, there is still

```
	at org.jvnet.hudson.test.RunLoadCounter$Marker.onLoad(RunLoadCounter.java:118)
	at hudson.model.Run.onLoad(Run.java:389)
	at hudson.model.RunMap.retrieve(RunMap.java:233)
	at hudson.model.RunMap.retrieve(RunMap.java:61)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:650)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:632)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:530)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.search(AbstractLazyLoadRunMap.java:495)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.newestBuild(AbstractLazyLoadRunMap.java:428)
	at jenkins.model.lazy.LazyBuildMixIn.getLastBuild(LazyBuildMixIn.java:254)
	at hudson.model.AbstractProject.getLastBuild(AbstractProject.java:943)
	at hudson.model.AbstractProject.getLastBuiltOn(AbstractProject.java:1016)
	at hudson.model.queue.MappingWorksheet$WorkChunk.<init>(MappingWorksheet.java:197)
	at hudson.model.queue.MappingWorksheet.<init>(MappingWorksheet.java:381)
	at hudson.model.queue.MappingWorksheet.<init>(MappingWorksheet.java:313)
	at hudson.model.Queue.maintain(Queue.java:1650)
```

Permalink updates, dashboard views, and extension points in various plugins can further force `lastBuild` to be loaded, so here I am only striving to avoid loading the builds prior to that. There is also #1191 which causes (seen here in `workflow-job`):

```
	at org.jvnet.hudson.test.RunLoadCounter$Marker.onLoad(RunLoadCounter.java:118)
	at hudson.model.Run.onLoad(Run.java:389)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.onLoad(WorkflowRun.java:566)
	at hudson.model.RunMap.retrieve(RunMap.java:233)
	at hudson.model.RunMap.retrieve(RunMap.java:61)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:650)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.load(AbstractLazyLoadRunMap.java:632)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.getByNumber(AbstractLazyLoadRunMap.java:530)
	at jenkins.model.lazy.AbstractLazyLoadRunMap.search(AbstractLazyLoadRunMap.java:495)
	at jenkins.model.lazy.LazyBuildMixIn$RunMixIn.getPreviousBuild(LazyBuildMixIn.java:400)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getPreviousBuild(WorkflowRun.java:273)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.getPreviousBuild(WorkflowRun.java:136)
	at jenkins.model.lazy.LazyBuildMixIn.newBuild(LazyBuildMixIn.java:194)
	at jenkins.model.ParameterizedJobMixIn$ParameterizedJob.createExecutable(ParameterizedJobMixIn.java:505)
	at jenkins.model.ParameterizedJobMixIn$ParameterizedJob.createExecutable(ParameterizedJobMixIn.java:323)
	at hudson.model.Executor$1.call(Executor.java:370)
	at hudson.model.Executor$1.call(Executor.java:352)
	at hudson.model.Queue._withLock(Queue.java:1456)
	at hudson.model.Queue.withLock(Queue.java:1312)
	at hudson.model.Executor.run(Executor.java:352)
```

On the topic of `MappingWorksheet`, it is worth noting that https://github.com/jenkinsci/jenkins/blob/c329f2254959e16efa6b3ecb1036d7adfd5b88ee/core/src/main/java/hudson/model/queue/MappingWorksheet.java#L327-L356 including `LoadPredictor` and `FutureLoad` exists solely for the benefit of https://github.com/jenkinsci/slave-squatter-plugin which [seems dead](https://issues.jenkins.io/issues/?jql=component%3D16076) since 2016 or so and has this summary which seems to come from another generation:

> It is useful for making Jenkins aware of activities outside Jenkins — such as when you are using people's workstations as slaves (which need to be available for interactive use during day time)

### Proposed changelog entries

- Reduce the circumstances under which recent old builds will be loaded when starting new builds.

### Proposed upgrade guidelines

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7998"></a>

### CODE REVIEW

File: test/pom.xml
Comment (jglick) [inline test/pom.xml:109]: https://github.com/jenkinsci/jenkins-test-harness/pull/583

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Review by Vlatombe — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:282]: @jglick am I missing something here?
`Result threshold : List.of(Result.UNSTABLE, Result.FAILURE` 

if there are only 2 loaded builds (both success) (or 5 loaded builds but only 2 of which are successful)  won't this cause the loop to add both of them for the `UNSTABLE` case and then add one duplicate for the FAILURE case?

Should the `List<RunT>` be a `Set` instead and only switched to a List at the end.  (not sure why a list and not a `SortedSet` - as duplicate Runs don;t make sense here).

Comment (jglick) [reply core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:282]: Possibly. I will check soon if there is existing test coverage specifically for this method, and see if I can reproduce any issues like that.

Comment (jglick) [reply core/src/main/java/jenkins/model/lazy/LazyBuildMixIn.java:282]: #8233. Sorry for the delay.

**Comment by jglick** [comment]
Just noticed I had filed this years earlier: https://issues.jenkins.io/browse/JENKINS-50030

**Pull request merged**
